### PR TITLE
writebuffers: Temp filesystem support

### PR DESCRIPTION
**Pull request by frznlogic**
This commit adds support for a tmpfs with a fixed size to be added on the temporary filesystem created for creating all container write buffers on. This will limit the impact of writes and make them contained to separate filesystems completely.

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

### CODE REVIEW

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (kursatkobya) [inline agent/src/containeroptions/containeroptionparser.cpp]: this should not be bool i guess

Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: no it should not ;)

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (kursatkobya) [inline agent/src/containeroptions/containeroptionparser.cpp]: so basically this is a default temporaryFileSystemSize. 
So i would prefer to have this value as a constexpr in class

Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: It can't be a constexpr as it's set to a value on the line below this comment. The value here is just a "sane default" in case JSONParser fails for some reason.

Comment (kursatkobya) [reply agent/src/containeroptions/containeroptionparser.cpp]: But it can be something like 
`int temporaryFileSystemSize = DEFAULT_TEMP_FSIZE;`
right

Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: Ahh, I see, sorry I completely misunderstood you =).

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (kursatkobya) [inline agent/src/containeroptions/containeroptionparser.cpp]: why don't we directly log_output the message like
log_warn() << "Could not parse config due to: 'enableTemporaryFilesystemWriteBuffer' not found.";

Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: It was used for a DBusException before as well, I'll fix it.

**Review by kursatkobya — APPROVED** [review]
looks good

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (kursatkobya) [inline agent/src/containeroptions/containeroptionparser.cpp]: it would be nice if you can find better name instead wo

Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: Done.

File: agent/src/containeroptions/dynamiccontaineroptions.cpp
Comment (thenor) [inline agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: Shouldn't this be called something like "isTemporaryFile...Enabled"?

Comment (frznlogic) [reply agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: it's a getter for the parameter named in the same manner as the other getters though?

Comment (thenor) [reply agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: I don't know if we have any code guidelines for this, but I am used to that getters for bools are "isVariable" and getters for other types are "getVariable". Not sure if we have had or should have that in SC.

File: agent/src/containeroptions/dynamiccontaineroptions.cpp
Comment (thenor) [inline agent/src/containeroptions/dynamiccontaineroptions.cpp:58]: As my previous statement says I would prefer if this is called "getTemporary..."

File: agent/src/containeroptions/dynamiccontaineroptions.cpp
Comment (thenor) [reply agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: It's just that I find this method name confusing since it doesn't enable anything. But I understand that the variables name is "enable...".

File: agent/unit-test/containeroptionparser_unittest.cpp
Comment (thenor) [inline agent/unit-test/containeroptionparser_unittest.cpp]: I would prefer if the names of the test cases were something a little more descriptive. Also for the first and second of the "Nice" test cases (and "Bad"). Just a preference. It makes it so much easier to understand what's going on in the test case without reading the full log.

File: doc/chapters/filesystem/index.rst
Comment (thenor) [inline doc/chapters/filesystem/index.rst]: Is there a reason for the upper case 'w' in "Write"?

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (breakreturn) [inline agent/src/containeroptions/containeroptionparser.cpp]: Move this up one line to after the if

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (breakreturn) [inline agent/src/containeroptions/containeroptionparser.cpp]: Move the { to the next line.

File: agent/src/containeroptions/dynamiccontaineroptions.cpp
Comment (frznlogic) [reply agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: I understand your point ;) . We do have a lot of Qt coding style guidelines which I think are reasonable, and this naming does follow those ... or well, they don't mention it really, but their code is the same as this at least ;). Not sure what others feel like? @tobsan @breakreturn ?

File: doc/chapters/filesystem/index.rst
Comment (frznlogic) [reply doc/chapters/filesystem/index.rst]: Nope. fixed.

File: agent/src/containeroptions/dynamiccontaineroptions.cpp
Comment (breakreturn) [reply agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: I don't think there are any guidelines. I'm also guessing we are mixing a bit, but the Qt style I believe is to have setX(), and just x() to get a property. A true/false check I do think is common to have isX() style though. I'd propose that @frznlogic simply do as how it looks in the code around here, and if there are no good references, check the rest of the code base.

File: doc/chapters/filesystem/index.rst
Comment (thenor) [inline doc/chapters/filesystem/index.rst:167]: I can't see that you have a test case for when the flag is set to false and no size is included.

File: agent/unit-test/containeroptionparser_unittest.cpp
Comment (frznlogic) [reply agent/unit-test/containeroptionparser_unittest.cpp]: Fair enough. Fixed naming.

Comment (breakreturn) [reply agent/unit-test/containeroptionparser_unittest.cpp]: On a glance, this looks like a good candidate for parametrization. If not, then I definitely think they should have proper names and comments on each of them what's being tested.

File: doc/chapters/filesystem/index.rst
Comment (thenor) [inline doc/chapters/filesystem/index.rst]: Is the ' at the end of the line a typo? Is it supposed to mean plural?

File: doc/chapters/filesystem/index.rst
Comment (frznlogic) [reply doc/chapters/filesystem/index.rst:167]: There was not. I added a test now, thanks!

File: doc/chapters/filesystem/index.rst
Comment (breakreturn) [inline doc/chapters/filesystem/index.rst]: Should "Create DBus" really by code highlighted? I guess you only want the method name to be that?

File: doc/chapters/filesystem/index.rst
Comment (frznlogic) [reply doc/chapters/filesystem/index.rst]: Weird, it's supposed to be highlighted... it's meant to be in plural though.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: What's expected?

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: It's "SoftwareContainerAgent". And this comment doesn't really say what's going to be tested either, what's the expected behavior?

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: Why is this one test? It looks like two separate tests in one.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: What's that number? Is the exact number significant? I think it's better to have a variable.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: Where is this exception asserted?

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: Fixed.

File: doc/chapters/filesystem/index.rst
Comment (frznlogic) [reply doc/chapters/filesystem/index.rst]: Fixed.

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: Files larger than 10MB can not be written and files smaller than 9MB can be written :). I'll add it.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: What's this number?

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py]: What are these number and the ones below?

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: servicetest/filesystem/test_filesystem.py
Comment (thenor) [inline servicetest/filesystem/test_filesystem.py]: Where do you check for the DBus exception?

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: Added/fixed and renamed the test.

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: I changed the behavior during writing the tests and forgot to update the comments. Sorry, it's fixed now.

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: No good reason to be honest except trying to shorten down runtime. I'll break it out.

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: it's 10 \* 1024 \* 1024 or 10 megabytes.

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: it's 10 \* 1024 \* 1024

File: servicetest/filesystem/test_filesystem.py
Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: It's 9 megabytes. Just a random number =)

Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: Comments have been added.

File: servicetest/filesystem/test_filesystem.py
Comment (thenor) [inline servicetest/filesystem/test_filesystem.py]: Is it guaranteed that the ID for the container will be "SC-0"? I think there are other test cases that fetch the ID instead of hardcoding it.

File: servicetest/filesystem/test_filesystem.py
Comment (thenor) [inline servicetest/filesystem/test_filesystem.py]: Will this also verify that it was successful? Will this fail if it is not possible?

**Review by thenor — CHANGES_REQUESTED** [review]
My comments on the code are not blocking the merge to master, in my opinion, but the comments on the service tests I would like to get answers for before approval.

File: servicetest/filesystem/test_filesystem.py
Comment (kursatkobya) [reply servicetest/filesystem/test_filesystem.py]: service tests for cgroups had that i guess

Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py]: It's a good idea. That said, we have similar problems in a lot of places in the testing code, but that's no reason to make it worse i guess ;). Fixed.

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (breakreturn) [inline agent/src/containeroptions/containeroptionparser.cpp]: The '{' goes on the next line when the if is broken up like this, as I mentioned in previous review.

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (breakreturn) [inline agent/src/containeroptions/containeroptionparser.cpp]: Same here with the {

File: agent/unit-test/containeroptionparser_unittest.cpp
Comment (breakreturn) [inline agent/unit-test/containeroptionparser_unittest.cpp]: I think all tests should have a comment what they test, what the values mean, what the purpose is etc. Please comment the tests. If it's deemed overkill to have a comment on each, unlikely IMO, then at least a comment before the batch of functions that explains the overall strategy.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [inline servicetest/filesystem/test_filesystem.py:387]: Again, why not have descriptive names instead of values like this?

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: agent/src/containeroptions/containeroptionparser.cpp
Comment (frznlogic) [reply agent/src/containeroptions/containeroptionparser.cpp]: Fixed. However, these rules are counterintuitive imho. I'd prefer if we made them more intuitive.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [reply servicetest/filesystem/test_filesystem.py:387]: @frznlogic In this diff I actually missed the comment above, sorry! I think this is explanatory enough, the "magic" numbers aren't magic anymore with the above comments. Sorry.

Comment (frznlogic) [reply servicetest/filesystem/test_filesystem.py:387]: The comments above where added to alleviate the understanding of the values. I hope those are enough?

File: agent/src/containeroptions/dynamiccontaineroptions.cpp
Comment (frznlogic) [reply agent/src/containeroptions/dynamiccontaineroptions.cpp:48]: I'm staying with the Qt style as discussed at todays meetings but will look into renaming the parameters.

File: servicetest/filesystem/test_filesystem.py
Comment (breakreturn) [reply servicetest/filesystem/test_filesystem.py:387]: Yes it looks perfectly fine to me.

**Review by breakreturn — APPROVED** [review]

**Review by thenor — APPROVED** [review]

**Pull request merged**
